patroni/dcs/__init__.py

```diff
@@ -1042,35 +1037,48 @@ def _get_permanent_slots(self, is_standby_cluster: bool, role: str, nofailover:
 
         return self.__permanent_slots if role in ('master', 'primary') else self.__permanent_logical_slots
 
-    def _get_slot_members(self, my_name: str, role: str) -> List[str]:
+    def _get_members_slots(self, my_name: str, role: str) -> Dict[str, Dict[str, str]]:
```

Comment:
I'm not convinced this makes sense. You are changing the meaning of this method, whilst making it do the same as the old, adding a new purpose.

I would either:
1. Create a new method called `_get_members_slots`, which to me means "get the slots belonging to a member", and reference this after the call to `_get_slot_members`. This makes sure that `_get_members_slots` follows the "do one thing, well" principle.
2. (and honestly I much prefer 1 :smile:) update the docstring so that it better describe the new role of this method.

Just to expand on 1, I think that following the "one thing" rule will help to keep the complexity to a reasonable level whilst encouraging a flatter structure. It makes it easier to follow code, without needing to read the depth of it. Docstrings become easier to write. Method signatures need fewer args. Modules can be shorter...

Reply:
There are very good reasons why I changed the method.
I am planning to create physical replication slots on all members that could become the leader and use pg_replication_slot_advance() function to move them forward based on data from `/member/*` keys.

For that I need to add `lsn` value to the `{'type': 'physical'}` (similarly to https://github.com/zalando/patroni/pull/2851/files#diff-247edd60473c0933e528a1b4621cc612da54b6dc964720719d7623135d312474R914). Doing it in the get_replication_slots() method seems totally wrong, therefore I renamed _get_slot_members() to _get_members_slots() and moved piece of the code that builds a structure there. But, forgot to update `Get a list of member names that have replication slots sourcing from this node` docstring.

Reply:
Thanks for the extra info. So you're intending on using `_get_members_slots()` from other methods?

I didn't mean to suggest that you didn't have good reason to make the change. Apologies if that's how it came across. I wasn't saying don't move the code out of `__permanent_physical_slots`, I meant have 2 separate methods for each of "getting a list of members" and "getting the slots of a member". If you think a separate `_get_slot_members` is overkill as they will always be together, that's good.

Updating the docstring is good too. :smile:

Reply:
`_get_members_slots()` will be called as before, only from `get_replication_slots()`.

The new method isn't big and IMO it is way to early for splitting it.